bluepyopt/ephys/simulators.py

```diff
@@ -182,3 +182,161 @@ def __init__(self, message, original):
 
         super(NrnSimulatorException, self).__init__(message)
         self.original = original
+
+
+class LFPySimulator(object):
+
+    """LFPy simulator"""
+
+    def __init__(self, LFPyCellModel, electrode=None, cvode_active=True,
+                 cvode_minstep=None, random123_globalindex=None,
+                 mechanisms_directory=None):
+        """Constructor
+        Args:
+            LFPyCellModel (LFPyCellModel): the LFPy cell model
+            electrode (MEAutility.MEA): the MEAutility probe used to compute
+                extracellular signals
+            cvode_active (bool): should neuron use the variable time step
+                integration method
+            cvode_minstep (float): the minimum time step allowed for a cvode
+                step. Default is 0.0.
+            random123_globalindex (int): used to set the global index used by
+                all instances of the Random123 instances of Random
+            mechanisms_directory (str): path to the parent directory of the
+                directory containing the mod files. If the mod files are in
+                "./data/mechanisms", then mechanisms_directory should be
+                "./data/".
+        """
+
+        self.LFPyCellModel = LFPyCellModel
```

Comment:
I find it weird that the Simulator has a model inside

Reply:
that is required by LFPy since the `simulate` fucntion is from the `LFPy.Cell` object:  https://github.com/LFPy/LFPy/blob/master/LFPy/cell.py#L992

Reply:
but do you already need the simulate function here? Isn't that something for in the simulate below?

Reply:
I see, so you suggest moving the `lfpycell` and `electrode` arguments directly to the `run` function.

Reply:
actrually, then we'd move the problem to the [protocol.run()](https://github.com/alejoe91/BluePyOpt/blob/lfpy/bluepyopt/ephys/protocols.py#L189) function and we would need additional parameters there to be propagated everywhere.

Reply:
Something like:
```

class LFPySimulator(NrnSimulator):

    """LFPy simulator"""

    def __init__(self, dt=None, cvode_active=True, cvode_minstep=None,
                 random123_globalindex=None, mechanisms_directory=None):
        """Constructor

        Args:
            dt (float): the integration time step used by neuron.
            cvode_active (bool): should neuron use the variable time step
                integration method
            cvode_minstep (float): the minimum time step allowed for a cvode
                step. Default is 0.0.
            random123_globalindex (int): used to set the global index used by
                all instances of the Random123 instances of Random
            mechanisms_directory (str): path to the parent directory of the
                directory containing the mod files. If the mod files are in
                "./data/mechanisms", then mechanisms_directory should be
                "./data/".
        """

        super(LFPySimulator, self).__init__(
            dt=dt,
            cvode_active=cvode_active,
            cvode_minstep=cvode_minstep,
            random123_globalindex=random123_globalindex,
            mechanisms_directory=mechanisms_directory
        )

    def run(
            self,
            lfpy_cell,
            lfpy_electrode,
            tstop=None,
            dt=None,
            cvode_active=None,
            random123_globalindex=None):
        """Run protocol"""
        import LFPy
        _ = self.neuron

        lfpy_cell.tstart = 0.0
        lfpy_cell.tstop = tstop

        if dt is not None:
            lfpy_cell.LFPyCell.dt = dt

        if cvode_active and dt is not None:
            raise ValueError(
                'NrnSimulator: Impossible to combine the dt argument when '
                'cvode_active is True in the NrnSimulator run method')

        if cvode_active is None:
            cvode_active = self.cvode_active

        if cvode_active is not None:
            self.cvode_active = cvode_active

        if random123_globalindex is None:
            random123_globalindex = self.random123_globalindex

        if random123_globalindex is not None:
            rng = self.neuron.h.Random()
            rng.Random123_globalindex(random123_globalindex)

        probes = [lfpy_electrode] if lfpy_electrode is not None else None

        sim_params = {
            "probes": probes,
            "rec_vmem": False,
            "rec_imem": False,
            "rec_ipas": False,
            "rec_icap": False,
            "rec_variables": [],
            "variable_dt": self.cvode_active,
            "atol": 0.001,
            "to_memory": True,
            "to_file": False,
            "file_name": None,
        }

        try:
            lfpy_cell.simulate(**sim_params)
        except Exception as e:
            raise LFPySimulatorException("LFPy simulator error", e)

        logger.debug("LFPy simulation finished")

```